x/dsmr/partition.go

```diff
@@ -0,0 +1,217 @@
+	for _, vdr := range vdrs {
+		weightedValidators = append(weightedValidators, &weightedValidator{
+			weight: vdr.Weight,
+			nodeID: vdr.NodeID,
+		})
+	}
+	return weightedValidators, nil
+}
+
+func precomputePartition[T Tx](validators []*weightedValidator) *PrecalculatedPartition[T] {
+	utils.Sort(validators)
+	accumulatedWeight := uint64(0)
+	for _, weightedVdr := range validators {
+		accumulatedWeight += weightedVdr.weight
+		weightedVdr.accumulatedWeight = accumulatedWeight
+	}
+	return &PrecalculatedPartition[T]{
+		validators:  validators,
+		totalWeight: accumulatedWeight,
+	}
+}
+
+func CalculatePartition[T Tx](ctx context.Context, state validators.State, pChainHeight uint64, subnetID ids.ID) (*PrecalculatedPartition[T], error) {
+	weightedValidators, err := getWeightedVdrsFromState(ctx, state, pChainHeight, subnetID)
+	if err != nil {
+		return nil, err
+	}
+
+	return precomputePartition[T](weightedValidators), nil
+}
+
+func calculateSponsorWeightIndex(sponsor codec.Address, totalWeight uint64) uint64 {
+	return binary.BigEndian.Uint64(sponsor[len(sponsor)-consts.Uint64Len:]) % totalWeight
+}
+
+func (pp *PrecalculatedPartition[T]) AssignTx(tx T) (ids.NodeID, bool) {
+	if pp.totalWeight == 0 || len(pp.validators) == 0 {
+		return ids.NodeID{}, false
+	}
+
+	sponsor := tx.GetSponsor()
+	sponsorWeightIndex := calculateSponsorWeightIndex(sponsor, pp.totalWeight)
+	nodeIDIndex := sort.Search(len(pp.validators), func(i int) bool {
+		return pp.validators[i].accumulatedWeight > sponsorWeightIndex
+	})
+
+	// Defensive: this should never happen
+	if nodeIDIndex >= len(pp.validators) {
+		return ids.NodeID{}, false
+	}
+	return pp.validators[nodeIDIndex].nodeID, true
+}
+
+type indexedTx[T Tx] struct {
+	index uint64
+	tx    T
+}
+
+func (i *indexedTx[T]) Compare(o *indexedTx[T]) int {
+	switch {
+	case i.index < o.index:
+		return -1
+	case i.index > o.index:
+		return 1
+	default:
+		return 0
+	}
+}
+
+func (pp *PrecalculatedPartition[T]) calculateIndexedTxs(txs []T) []*indexedTx[T] {
+	indexedTxs := make([]*indexedTx[T], len(txs))
+	for i, tx := range txs {
+		sponsor := tx.GetSponsor()
+		sponsorWeightIndex := calculateSponsorWeightIndex(sponsor, pp.totalWeight)
+		indexedTxs[i] = &indexedTx[T]{
+			index: sponsorWeightIndex,
+			tx:    tx,
+		}
+	}
+	utils.Sort(indexedTxs)
+	return indexedTxs
+}
+
+func (pp *PrecalculatedPartition[T]) AssignTxs(txs []T) (map[ids.NodeID][]T, error) {
+	indexedTxs := pp.calculateIndexedTxs(txs)
+
+	assignments := make(map[ids.NodeID][]T)
+	weightIndex := uint64(0)
+	txIndex := 0
+	for _, vdr := range pp.validators {
+		weightIndex += vdr.weight
+		for txIndex < len(indexedTxs) && weightIndex > indexedTxs[txIndex].index {
+			assignments[vdr.nodeID] = append(assignments[vdr.nodeID], indexedTxs[txIndex].tx)
+			txIndex++
+		}
+		if txIndex >= len(indexedTxs) {
+			break
+		}
+	}
+	return assignments, nil
+}
+
+func (pp *PrecalculatedPartition[T]) FilterTxs(nodeID ids.NodeID, txs []T) ([]T, error) {
```

Comment:
Can't we also get rid of this function as well? Seems like filter could just be implemented by calling `AssignTx` in a loop over `txs` and just filtering out anything not returning `nodeID`.

Reply:
Removed